kafka-client/src/main/java/dev/responsive/kafka/internal/stores/SizeTrackingBuffer.java

```diff
@@ -32,7 +33,7 @@ public class SizeTrackingBuffer<K extends Comparable<K>> {
 
   public SizeTrackingBuffer(final KeySpec<K> extractor) {
     this.extractor = Objects.requireNonNull(extractor);
-    buffer = new TreeMap<>();
+    buffer = new ConcurrentSkipListMap<>();
```

Comment:
While I think it should be fine given the frequency of flushing, this data structure can start to break down at scale (in terms of number of unique keys) and using it for the in-memory KV store was the reason behind a major performance regression in C3/Streams a long time ago.

Of course, that would only grow in size whereas the CommitBuffer would only accumulate as many unique keys as are seen within the commit triggers. Still, we should consider running a benchmark and/or doing some quick math on what this might look like for a deduper-like app that contains mostly new keys

Reply:
Ah, one more thing I remembered is that this doesn't have an O(1) `size()` implementation (for some wild reason) so if we ever call `#size` on this map, we should use an AtomicInt to keep track of the size ourselves instead

IIRC `isEmpty` is fine, it's just `size` that needs to be avoided

Reply:
Noted - I was not aware of the perf issues with it. I'm hopeful that we'll be ok because (1) like you said, the commit buffer size is relatively small and is bounded, and (2) our workloads tend to not be cpu-bound, but instead bound on i/o to the stores. Given that, I'm leaning toward trying to keep the implementation here simple and keep the current data structure. We'll see if it's problematic when we benchmark the async stuff.